File: source/swift/src/Tarjan.swift
// Tarjan's Union-Find Algorithm

public class Tarjan {
    public static let name = "Tarjan"
    private var root: [Int]
    private(set) var sets: Int

    public init(n: Int) {
        root = Array(0..<n)
        sets = n
    }

    public func reset(n: Int) {
        sets = n
        for i in 0..<n {
            root[i] = i
        }
    }

    public func find(_ a: Int) -> Int {
        var here = a

        while root[here] != here {
            here = root[here]
        }
        let top = here

        here = a
        while root[here] != top {
            let next = root[here]
            root[here] = top
            here = next
        }

        return top
    }

    public func unite(_ a: Int, _ b: Int) {
        let a = find(a)
        let b = find(b)

        if a != b {
            sets -= 1
            root[a] = b
        }
    }

    public func setCount() -> Int {
        return sets
    }
}

File: source/rust/src/tarjan.rs
// Tarjan's Union-Find Algorithm

#[allow(dead_code)]
pub(crate) struct Tarjan {
    seed: Vec<usize>,
    root: Vec<usize>,
    sets: usize,
}

#[allow(dead_code)]
impl Tarjan {
    pub(crate) const NAME: &'static str = "Tarjan";

    pub(crate) fn create(n: usize) -> Self {
        let seed: Vec<usize> = (0..n).collect();
        let root = vec![0; n];
        Tarjan { seed, root, sets: n }
    }

    pub(crate) fn reset(&mut self, n: usize) {
        self.root[..n].copy_from_slice(&self.seed[..n]);
        self.sets = n;
    }

    fn find(&mut self, a: usize) -> usize {
        let mut here = a;

        while self.root[here] != here {
            here = self.root[here];
        }
        let top = here;

        here = a;
        while self.root[here] != top {
            let next = self.root[here];
            self.root[here] = top;
            here = next;
        }

        top
    }

    pub(crate) fn unite(&mut self, a: usize, b: usize) {
        let a = self.find(a);
        let b = self.find(b);

        if a != b {
            self.sets -= 1;
            self.root[a] = b;
        }
    }

    pub(crate) fn set_count(&self) -> usize {
        self.sets
    }
}
